mmeds/util.py

```diff
@@ -1125,27 +1131,166 @@ def parse_barcodes(forward_barcodes, reverse_barcodes, forward_mapcodes, reverse
+
+    return ret_val
+
+
+def validate_demultiplex(demux_file, for_barcodes, rev_barcodes, map_file, log_dir, is_gzip, get_read_counts=False,
+                         on_chimera=True):
+    """
+    Calls qiime1 script to validate a gzipped, demultiplex fastq file.
+    source_dir: where the data is located.
+    demux_file: path to gzipped, demultiplexed, fastq file.
+    for_barcodes: path to gzipped, forward barcode file
+    rev_barcodes: path to reverse barcode file
+    map_file: path to a qiime mapping file
+    log_dir: path to log dir, where the results will be saved
+    """
+    # TODO: generalize for single barcodes
+
+    # Allows us to test that the correct barcodes are in the resulting demultiplexed file.
+    barcode_return = create_barcode_mapfile(Path(demux_file).parent, for_barcodes, rev_barcodes,
+                                            Path(demux_file).stem, map_file, get_read_counts)
+    if get_read_counts:
+        map_df = barcode_return[0]
+        matched_barcodes = barcode_return[1]
+        all_barcodes = barcode_return[2]
+    else:
+        map_df = barcode_return
+
+    new_env = setup_environment('qiime/1.9.1')
+
+    test_file = f'{Path(demux_file).parent}/{Path(demux_file).stem}_test.fastq'
+    map_file = f'{Path(demux_file).parent}/{Path(demux_file).stem}_qiime_barcode_mapfile.tsv'
+
+    gunzip_demux_file = ['gunzip', f'{demux_file}.gz']
+    gzip_demux_file = ['gzip', demux_file]
+    create_fastq_copy = ['cp', f'{demux_file}', f'{test_file}']
+
+    # to create the fasta file, on every line mod4 = 1, replace @ with >
+    # then on every line mod4 = 2, print the line. then skip all other lines.
+    create_fasta_file = ['sed', '-n', '-i', '1~4s/^@/>/p;2~4p', f'{test_file}']
+
+
+    # Call qiime1 validate demultiplex script
+    if on_chimera:
+        validate_demux_file = f'ml python/2.7.9-UCS4; validate_demultiplexed_fasta.py -b -a\
+                            -i {test_file}\
+                            -o {log_dir}\
+                            -m {map_file};\
+                            rm {test_file}'
+    else:
+        validate_demux_file = f'validate_demultiplexed_fasta.py -b -a\
+                            -i test_file\
+                            -o {log_dir}\
+                            -m {map_file};\
+                            rm {test_file}'
+
+    try:
+        if is_gzip:
+            run(gunzip_demux_file, capture_output=True, check=True)
+
+        run(create_fastq_copy, capture_output=True, check=True)
+        run(create_fasta_file, capture_output=True, check=True)
+        validate_output = run(validate_demux_file, capture_output=True, env=new_env, check=True, shell=True)
+
+        if is_gzip:
+            run(gzip_demux_file, capture_output=True, check=True)
+
+    except CalledProcessError as e:
+        Logger.debug(e)
+        print(e.output)
+        raise e
+
+    if get_read_counts:
+        ret_val = (validate_output, matched_barcodes, all_barcodes)
```

Comment:
i've warmed up to python a lot in the past few months, but it's gross that the return type isn't defined and you can just do this lol

Reply:
;) I'd rather do multiple returns but that isn't pythonic